### PR TITLE
[Merged by Bors] - feat: make alias_in work with the module system

### PR DESCRIPTION
**Pull request by scholzhannah**
This requires two different fixes:
- deal with `private` aliases: under the module system, a private declaration `Foo.Bar` in module `MyMod` gets a unique name (currently something like `_private.MyMod.Foo.Bar`). This breaks the current logic for counting namespace components. Since we don't see a use for private aliases (they cannot export anything to other modules), we make the attribute error on them.
- for some reason, we have to create `public alias` inside the attribute.

---

I finally have time now to actually make a PR using this attribute for CW complexes but it needs to work with the module system first. Since for private declarations, the namespace is longer and thus the previous approach did not work, the easiest fix was to simply not allow this attribute to be used for private declarations. I don't know why anyone would create private aliases, so I hope this is fine. 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by scholzhannah** [comment]
Should I still add a follow-up PR with a module deprecation even though this is just a test?

File: Mathlib/Util/AliasIn.lean [hunk 1]
```diff
@@ -50,12 +51,15 @@ initialize registerBuiltinAttribute {
       let newNamespace := nm.getId.components
       let num := num.map (·.getNat) |>.getD newNamespace.length
       let components := src.components
+      if (← toModifiers src).visibility.isPrivate then
+        throwError "The `alias_in` attribute cannot be used for private declarations."
```
Comment (grunweg) [inline Mathlib/Util/AliasIn.lean:55]:
```suggestion
        throwError "the `alias_in` attribute cannot be used for private declarations."
```

File: MathlibTest/Util/AliasIn/AliasInModuleSystem.lean [hunk 2]
```diff
@@ -1,4 +1,8 @@
-import Mathlib.Util.AliasIn
+module
+
+public import Mathlib.Util.AliasIn
+
```
Comment (grunweg) [inline MathlibTest/Util/AliasIn/AliasInModuleSystem.lean:4]:
```suggestion

/-! Tests for the `@[alias_in]` attribute, within the module system  -/

```

File: MathlibTest/Util/AliasIn/AliasInModuleSystem.lean [hunk 3]
```diff
@@ -64,3 +68,11 @@ Look at this docstring!
 #guard_msgs in
 open Lean in
 run_cmd logInfo m!"{(← Lean.findDocString? (← getEnv) `Foo.Baz.baz5).get!}"
+
```
Comment (grunweg) [inline MathlibTest/Util/AliasIn/AliasInModuleSystem.lean:73]: Let's add a test for explicit `private` inside the a public section.
```suggestion

@[alias_in Baz] public def Foo.Bar.baz' : Nat := 1

/-- error: The `alias_in` attribute cannot be used for private declarations. -/
#guard_msgs in
@[alias_in Baz] private def Foo.Bar.baz'' : Nat := 1

```

File: MathlibTest/Util/AliasIn/AliasInModuleSystem.lean [hunk 4]
```diff
@@ -64,3 +68,11 @@ Look at this docstring!
...
+end
+
+section
+
```
Comment (grunweg) [inline MathlibTest/Util/AliasIn/AliasInModuleSystem.lean:83]: And another for explicit visibilities inside a non-public section.
```suggestion

#guard_msgs in
@[alias_in Baz] public def Foo.Bar.n := 2

/-- error: The `alias_in` attribute cannot be used for private declarations. -/
#guard_msgs in
@[alias_in Baz] private def Foo.Bar.n' := 2

```

File: MathlibTest/Util/AliasIn/AliasIn.lean [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import Mathlib.Util.AliasIn
+
```
Comment (grunweg) [inline MathlibTest/Util/AliasIn/AliasIn.lean:2]: Let's add a comment about this file, such as:
```suggestion

/-! Tests for the `@[alias_in]` attribute without the module system -/

File: MathlibTest/Util/AliasIn/AliasIn.lean [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+/-- info: Foo.Baz.Qux.baz2 : Nat -/
+#guard_msgs in
+#check Foo.Baz.Qux.baz2
+
+/--
+error: Foo.Bar.baz3 has only 2 namespaces, cannot remove 3.
+Use `@[alias_in Baz.Qux 2]` instead.
+-/
+#guard_msgs in
+@[alias_in Baz.Qux 3] def Foo.Bar.baz3 : Nat := 2
```
Comment (grunweg) [inline MathlibTest/Util/AliasIn/AliasIn.lean:20]: Let's add a test about private definitions also:
```
/-- error: The `alias_in` attribute cannot be used for private declarations. -/
#guard_msgs in
@[alias_in Baz] private def Foo.Bar.baz5'' : Nat := 2

File: Mathlib/Util/AliasIn.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/Util/AliasIn.lean:55]: (and you'll need to update the tests accordingly)

**Comment by grunweg** [comment]
I think there's no need for that. (The point of the module deprecation is to not break imports of mathlib modules; mathlib tests are something else.)

**Review by grunweg — COMMENTED** [review]
A few minor comments. I tried to update the PR description to describe what this PR is actually doing. I realise I don't know either why the *public* is necessary. Let's find out!

File: Mathlib/Util/AliasIn.lean (hunk 1 above)
Comment (scholzhannah) [reply Mathlib/Util/AliasIn.lean:55]: Huh, why do we want this to be lowercase? As far as I can tell all the error messages always start in uppercase.

Comment (grunweg) [reply Mathlib/Util/AliasIn.lean:55]: The full error message printed is `error: The `alias_in` attribute cannot be used for private declarations.`,
and in my book, that's simply very odd capitalisation. (But I don't feel very strongly about this.)

Comment (scholzhannah) [reply Mathlib/Util/AliasIn.lean:55]: I think the `error: ` part just comes from `#guard_msgs`. I think if anything the capitalization is off there so I won't change it here.

Comment (grunweg) [reply Mathlib/Util/AliasIn.lean:55]: Let's get a second opinion :-)

**Review by grunweg — COMMENTED** [review]
This PR fixes an annoying bug, has tests covering all edge cases (I can think of) well, but I don't fully understand the second half of the fix. Perhaps another metaprogrammer can weigh in *why* the `public alias` is required now?
(Manually writing `alias` inside the file, in a `public section`, suffices... somehow, such effects are not inherited for the alias. Is that a different bug in the alias_in attribute? If so, perhaps the right fix is different...)

maintainer merge?

File: Mathlib/Util/AliasIn.lean (hunk 1 above)
Comment (adomani) [reply Mathlib/Util/AliasIn.lean:55]: Grepping for `"error: [a-z]"` yields 199 hits, while `"error: [A-Z]"` yields 156, so the odds seem to be slightly in favour of lowercase.

**Comment by adomani** [comment]
There seem to be 4 `private alias`es in mathlib:
```lean
Mathlib/Analysis/Complex/Exponential.lean:333:@[bound] private alias ⟨_, Bound.one_lt_exp_of_pos⟩ := one_lt_exp_iff
Mathlib/Data/ENNReal/Real.lean:178:@[bound] private alias ⟨_, Bound.ofReal_pos_of_pos⟩ := ofReal_pos
Mathlib/Data/NNReal/Defs.lean:325:@[bound] private alias ⟨_, Bound.coe_lt_coe_of_lt⟩ := coe_lt_coe
Mathlib/Data/NNReal/Defs.lean:330:@[bound] private alias ⟨_, Bound.coe_pos_of_pos⟩ := coe_pos
```
and they all have an attribute.  So, their use is very limited and probably intented to extend the `bound` tactic to work locally, but not globally.

**Comment by adomani** [comment]
> for some reason, we have to create `public alias` inside the attribute.

I also do not understand why this is needed.  Could you say what is the cause that you observed that made the `public` be necessary?

I think that `alias` is defined in Batteries, so maybe it is something about its implementation that needs to be addressed?  :shrug:

**Comment by scholzhannah** [comment]
> I also do not understand why this is needed. Could you say what is the cause that you observed that made the `public` be necessary?

If you remove the `public`, then for example the line `@[alias_in Baz] def Foo.Bar.baz : Nat := 1` in `AliasInModuleSystem.lean` gives the error `Unknown constant "Foo.Baz.baz"`. Note that writing 
```
def Foo.Bar.baz : Nat := 1

alias Foo.Baz.baz := Foo.Bar.baz
```
works perfectly, which in theory is just what this attribute is supposed to do.

**Comment by adomani** [comment]
Digging around a bit into this issue, it is the `Term.addTermInfo'` command that raises the error.

In fact, while at that point, the environment does not contain the new declaration, unless it is marked as public.

This seems unrelated to the earlier check that this PR also introduces, so it is probably just a consequence of the module system.

**Comment by adomani** [comment]
Thanks!  I am happy to go with this.  I'll let you decide whether to capitalise or not the error message.

If the `public` issue becomes annoying, we can reconsider.  However, the presence of the check that the attribute is applied to a `public` declaration avoids creating a `public alias` for a `private` declaration, which is definitely a good thing!

bors d+
